fish-rust/src/fish.rs

```diff
@@ -0,0 +1,877 @@
+    ];
+
+    let mut w = wgetopter_t::new(SHORT_OPTS, LONG_OPTS, args);
+    while let Some(c) = w.wgetopt_long() {
+        match c {
+            'c' => opts
+                .batch_cmds
+                .push(OsString::from_vec(wcs2string(w.woptarg.unwrap()))),
+            'C' => opts
+                .postconfig_cmds
+                .push(OsString::from_vec(wcs2string(w.woptarg.unwrap()))),
+            'd' => {
+                ffi::activate_flog_categories_by_pattern(w.woptarg.unwrap());
+                activate_flog_categories_by_pattern(w.woptarg.unwrap());
+                for cat in flog::categories::all_categories() {
+                    if cat.enabled.load(Ordering::Relaxed) {
+                        println!("Debug enabled for category: {}", cat.name);
+                    }
+                }
+            }
+            'o' => opts.debug_output = Some(OsString::from_vec(wcs2string(w.woptarg.unwrap()))),
+            'f' => opts.features = w.woptarg.unwrap().to_owned(),
+            'h' => opts.batch_cmds.push("__fish_print_help fish".into()),
+            'i' => opts.is_interactive_session = true,
+            'l' => opts.is_login = true,
+            'N' => {
+                opts.no_config = true;
+                // --no-config implies private mode, we won't be saving history
+                opts.enable_private_mode = true;
+            }
+            'n' => opts.no_exec = true,
+            RUSAGE_ARG => opts.print_rusage_self = true,
+            PRINT_DEBUG_CATEGORIES_ARG => {
+                let cats = flog::categories::all_categories();
+                // Compute width of longest name.
+                let mut name_width = 0;
+                for cat in cats.iter() {
+                    name_width = usize::max(name_width, cat.name.len());
+                }
+                // A little extra space.
+                name_width += 2;
+                for cat in cats.iter() {
+                    let desc = wgettext_str(cat.description);
+                    // https://doc.rust-lang.org/std/fmt/#syntax
+                    // this is left-justified
+                    println!("{:<width$} {}", cat.name, desc, width = name_width);
+                }
+                std::process::exit(0);
+            }
+            'p' => opts.profile_output = Some(OsString::from_vec(wcs2string(w.woptarg.unwrap()))),
+            PROFILE_STARTUP_ARG => {
+                // With "--profile-startup" we immediately turn profiling on.
+                opts.profile_startup_output =
+                    Some(OsString::from_vec(wcs2string(w.woptarg.unwrap())));
+                PROFILING_ACTIVE.store(true);
+                ffi::set_profiling_active(true);
+            }
+            'P' => opts.enable_private_mode = true,
+            'v' => {
+                // FIXME: this was _(L"%s, version %s\n"), but rust-fwprintf! takes a literal instead of an expr
+                // and appears to not print anything
```

Comment:
right so maybe that's why fwprintf doesn't do anything. You have my blessing to change `$format:literal` to `$format:expr` to support translated format strings. Either way, using `println` and `eprintln` is fine

Reply:
Okay I would like some clarification here:

As far as I understand it `fwprintf` in C++ _outputs_ in UTF32, my understanding of fish's UTF32-usage was that it is supposed to be an internal detail? Therefore changing it to `println!/eprintln!` makes it now print in UTF8. Is that a fine behaviour-change? As far as I can tell `fwprintf!` in Rust has only been used in `io.rs`, which I think is unadopted, so it might actually be non-working?

Reply:
yeah C++ uses wide output functions, while in Rust we always convert to bytes before printing.
I'm not sure what's the practical implications of this but I think I had cases where a mixture of the two would lead to output being dropped.
Maybe this is the case here as well? Not sure. If it causes problems, maybe replace it with a todo!, we'll probably not need it that much

Reply:
> As far as I understand it fwprintf in C++ outputs in UTF32,

It will output in the locale's encoding, which is typically UTF-8.

It will convert ("narrow") the arguments, or [in C++-speak](https://cplusplus.com/reference/cwchar/wprintf/):

"The external representation of wide characters in stdout are multibyte characters: These are obtained as if [wcrtomb](https://cplusplus.com/wcrtomb) was called to convert each wide character (using the stream's internal [mbstate_t](https://cplusplus.com/mbstate_t) object)."

Of course this will have awkward results if one of fish's internal private use codepoints gets out.

> I think I had cases where a mixture of the two would lead to output being dropped.

The case I know was that we mixed fwprintf (converting output) and fprintf ("direct" output), and with buffering turned on this would break, presumably because the two need separate buffers or need to have them flushed or something?

But it's important to know that `fwprintf(stderr, L"ä");` results in the bytes 0xc3 and 0xa4 (provided the locale is utf-8).

Reply:
Just to confirm, I have tried

```c++
std::fwprintf(stderr, L"%ls\n", L"\u00E4");
```

which results in output of 0xC3 0xA4 0x0A - which is utf-8 for `ä` (and a newline).

Reply:
I think I found my confusion: I was looking at [fwprintf](https://en.cppreference.com/w/c/io/fwprintf)[^1] then [std_streams](https://en.cppreference.com/w/c/io/std_streams)[^2], which lead me to [FILE](https://en.cppreference.com/w/c/io/FILE)

Which states:
> Narrow and wide orientation
A newly opened stream has no orientation. The first call to fwide or to any I/O function establishes the orientation: a wide I/O function makes the stream wide-oriented; a narrow I/O function makes the stream narrow-oriented. 

Which I then understood as `fwprintf` being a wide I/O function on `STDOUT`, and thus setting it to be wide-oriented **and thus outputting in wide**. But, I think my understanding of a `FILE`'s orientation is wrong, since that appears to only deal with what you send to it, not what it outputs, but cppreference.com does not make that distinction very clear.

It later states (in the middle of the same paragraph)

> Wide I/O functions convert between wide and multibyte characters as if by calling [std::mbrtowc](https://en.cppreference.com/w/cpp/string/multibyte/mbrtowc) or [std::wcrtomb](https://en.cppreference.com/w/cpp/string/multibyte/wcrtomb) with the conversion state as described by the stream.

Which I think is equivalent what cplusplus.com says about external representation, and is what I skipped.

For later: Is cppreference.com not a good place to look for C++ documentation? Should I prefer cplusplus.com?

It appears I mistakenly was on the C instead of C++ side of the documentation, but I don't think that should matter

[^1]: Which says nothing about the orientation/encoding/width
[^2]: Which says stdin/stderr is implicitly unoriented, whatever that means, turns out to be narrow/wide

Reply:
With that said: can I ask for a blessing to just use `println!/eprintln!` here? Using this: 

```rust
fwprintf!(libc::STDOUT_FILENO, "%s, version %s", PACKAGE_NAME, crate::BUILD_VERSION);
```

Prints nothing, and I don't want to debug `fwprintf!`. If we want to keep printf-style formatting I would rather do `println!("{}", sprintf!(...))`

Reply:
cppreference.com is the most useful one.
Yeah it's fine to use println.
It's weird because the Rust fwprintf does narrow IO. Maybe it fails to set the orientation or something.